cmd/red/proxy.go

```diff
@@ -0,0 +1,220 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"net/url"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/segmentio/conf"
+	consul "github.com/segmentio/consul-go"
+	"github.com/segmentio/events"
+	eventslog "github.com/segmentio/events/log"
+	redis "github.com/segmentio/redis-go"
+)
+
+type proxyConfig struct {
+	Bind     string `conf:"bind"     help:"Address on which the proxy is listening for incoming connections." validate:"nonzero"`
+	Upstream string `conf:"upstream" help:"URL or comma-separated list of upstream servers."                  validate:"nonzero"`
+	Debug    bool   `conf:"debug"    help:"Enable debug mode."`
+}
+
+func proxy(args []string) (err error) {
+	config := proxyConfig{
+		Bind: ":6479",
+	}
+
+	conf.LoadWith(&config, conf.Loader{
+		Name: "red proxy",
+		Args: args,
+		Sources: []conf.Source{
+			conf.NewEnvSource("RED", os.Environ()...),
+		},
+	})
+
+	events.DefaultLogger.EnableDebug = config.Debug
+	events.DefaultLogger.EnableSource = config.Debug
+
+	defer func() {
```

Comment:
It captures all panics that may be triggered by the code and converts them to an error which is returned by the function.